common/bucket_utils.py

```diff
@@ -0,0 +1,66 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Helper functions for interacting with the buckets."""
+
+from common import logs
+from common import experiment_utils
+
+logger = logs.Logger('bucket_utils')
+
+
+def _using_gsutil():
+    """ Test whether using gsutil format paths. """
+    try:
+        experiment_path_format = experiment_utils.get_cloud_experiment_path()
+    except KeyError:
+        experiment_path_format = 'gs://default'
+
+    return experiment_path_format.startswith('gs://')
+
+
+if _using_gsutil():
+    from common import gsutil as bucket_utils_impl
+else:
+    # When gsutil is not used in the context,
+    # here it should use local_utils.
+    # TODO(zhichengcai): local_utils
+    from common import gsutil as bucket_utils_impl
+
+
+def cp(*cp_arguments, **kwargs):  # pylint: disable=invalid-name
+    """ Copy source to destination. """
+    return bucket_utils_impl.cp(*cp_arguments, **kwargs)
+
+
+def ls(*ls_arguments, must_exist=True, **kwargs):  # pylint: disable=invalid-name
+    """ List files or folders. """
+    return bucket_utils_impl.ls(*ls_arguments, must_exist, **kwargs)
+
+
+def rm(*rm_arguments, recursive=True, force=False, **kwargs):  # pylint: disable=invalid-name
+    """ Remove files or folders. """
+    return bucket_utils_impl.rm(*rm_arguments, recursive, force, **kwargs)
+
+
+def rsync(  # pylint: disable=too-many-arguments
+        source,
+        destination,
+        delete=True,
+        recursive=True,
+        gsutil_options=None,
```

Comment:
If this is a generic rsync() function, should we remove gsutil_options?

Reply:
Right, I agree.
Currently, there is only one usages outside:

https://github.com/google/fuzzbench/blob/7554f6c9b090ee88d12cb531445ce1f936e6cd41/experiment/reporter.py#L45-L49

However, these options may not be available to other utils. For example, local_utils.
Then it will introduce problems later.

I have a plan:

Remove gsutil_options, instead, let options become a dictionary. When parsing options, we try to get the corresponded options to add. If corresponded options are empty, do nothing.

```
options={'gsutil': 
                  ['-h', 'Cache-Control:public,max-age=0,no-transform'] 
              }
```

What do you think?